workloads/librispeech/librispeech_jax/submission.py

```diff
@@ -0,0 +1,156 @@
+from jax import lax
+from workloads.librispeech.librispeech_jax.models import get_seq_lens
+
+
+def get_batch_size(workload_name):
+  batch_sizes = {"librispeech_jax": 8}
+  return batch_sizes[workload_name]
+
+def optimizer(hyperparameters: spec.Hyperparamters):
+  opt_init_fn, opt_update_fn = optax.adam(
+      learning_rate=hyperparameters.learning_rate
+    )
+  return opt_init_fn, opt_update_fn
+
+
+def init_optimizer_state(
+    workload: spec.Workload,
+    model_params: spec.ParameterContainer,
+    model_state: spec.ModelAuxiliaryState,
+    hyperparameters: spec.Hyperparamters,
+    rng: spec.RandomState) -> spec.OptimizerState:
+  params_zeros_like = jax.tree_map(
+      lambda s: jnp.zeros(s.shape_tuple), workload.param_shapes)
+  opt_init_fn, opt_update_fn = optimizer(hyperparameters)
+  optimizer_state = opt_init_fn(params_zeros_like)
+  return jax_utils.replicate(optimizer_state), opt_update_fn
+
+
+# We need to jax.pmap here instead of inside update_params because the latter
+# would recompile the function every step.
+@functools.partial(
+  jax.pmap,
+  axis_name='batch',
+  in_axes=(None, None, 0, 0, 0, None, 0, None),
+  static_broadcasted_argnums=(0, 1))
+def pmapped_train_step(workload, opt_update_fn, model_state, optimizer_state,
+                       current_param_container, hyperparameters, batch, rng):
+  def _loss_fn(params):
+    """loss function used for training."""
+    variables = {'params': params, **model_state}
+    logits, new_model_state = workload.model_fn(
+        params,
+        batch['input'],
+        model_state,
+        spec.ForwardPassMode.TRAIN,
+        rng,
+        update_batch_norm=True)
+    loss = workload.loss_fn(batch['label'], logits)
+    return loss, (new_model_state, logits)
+
+  grad_fn = jax.value_and_grad(_loss_fn, has_aux=True)
+  aux, grad = grad_fn(current_param_container)
+  grad = lax.pmean(grad, axis_name='batch')
+  new_model_state, logits = aux[1]
+  updates, new_optimizer_state = opt_update_fn(
+      grad, optimizer_state, current_param_container)
+  updated_params = optax.apply_updates(current_param_container, updates)
+
+  return new_model_state, new_optimizer_state, updated_params
+
+def update_params(
+    workload: spec.Workload,
+    current_param_container: spec.ParameterContainer,
+    current_params_types: spec.ParameterTypeTree,
+    model_state: spec.ModelAuxiliaryState,
+    hyperparameters: spec.Hyperparamters,
+    input_batch: spec.Tensor,
+    label_batch: spec.Tensor,
+    loss_type: spec.LossType,
+    optimizer_state: spec.OptimizerState,
+    eval_results: List[Tuple[int, float]],
+    global_step: int,
+    rng: spec.RandomState) -> spec.UpdateReturn:
+  """Return (updated_optimizer_state, updated_params, updated_model_state)."""
+  _, features, transcripts, input_lengths,  transcripts_padding = input_batch
+  features = jnp.expand_dims(features.transpose(0, 2, 1), axis=1)
+  num_devices = jax.local_device_count()
+  reshaped_features = jnp.reshape(
```

Comment:
I think that all of this reshaping, padding, etc. should be done in the input pipeline, because it should be the same regardless of the optimizer (and really `submission.py` should only define new code relating to the optimizer)